traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
-                if validate is not None:
-                    value = [validate(object, name, val) for val in value]
+        This simply calls the validator provided by the class, if any.
+        The validator is expected to have the signature::
 
-                list.__setitem__(self, slice(0, 0), value)
+            validator(trait_list, index, removed, added)
 
-                return
+        and return a validated added value or list of added values
+        or raise a TraitError.
 
-            except TraitError as excp:
-                excp.set_prefix("Each element of the")
-                raise excp
+        Parameters
+        ----------
+        index : int or slice
+            The indices being modified by the operation.
+        removed : list
+            The items to be removed.
+        added : list
+            The new items being added to the list.
 
-        self.len_error(len(value))
+        Returns
+        -------
+        values : list
+            The validated values that must be added to the list.
 
-    def _send_trait_items_event(self, name, event, items_event=None):
-        """ Send a TraitListEvent to the owning object if there is one.
+        Raises
+        ------
+        TraitError : Exception
+            If validation fails.
         """
-        object = self.object()
-        if object is not None:
-            if items_event is None and hasattr(self, "trait"):
-                items_event = self.trait.items_event()
-            object.trait_items_event(name, event, items_event)
+
+        # Use getattr as pickle can call `extend` before validator is set.
+        if getattr(self, 'validator', None) is None:
+            return added
+        else:
+            return self.validator(self, index, removed, added)
+
+    def notify(self, index, removed, added):
+        """ Call all notifiers
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(trait_list, index, removed, added)
+
+        Any return values are ignored.
+
+        Parameters
+        ----------
+        index : int or slice
+            The indices being modified by the operation.
+        removed : list
+            The items to be removed.
+        added : list
+            The items being added to the list.
+        """
+        # Use getattr as pickle can call `extend` before notifiers are set.
+        for notifier in getattr(self, 'notifiers', []):
+            notifier(self, index, removed, added)
+
+    def object(self):
+        """ Stub method to pass persistence tests. """
+        # XXX fix persistence tests to not introspect this!
+        return None
+
+    # ------------------------------------------------------------------------
+    # list interface
+    # ------------------------------------------------------------------------
+
+    def __init__(self, value=(), *, validator=None, notifiers=()):
```

Comment:
Nitpick: please could this constructor be moved to the top of the class? Partly because that's where code readers expect to find constructors, but partly because it's not actually part of the list interface, so doesn't really belong here.

Reply:
What's the reason for passing a tuple of `value` and `notifiers`? I'd expect those to be lists. For dealing with the usual mutable defaults problem, please use the standard approach of supplying `None` and then converting it appropriate in the body.

Reply:
Hmm; actually, in Python 3.8 the `list` type has signature: `list(iterable=(), /)`, so an empty tuple may be reasonable for the `value`.

Ideally we'd make `value` positional-only so that it can't be passed by name, but we have to wait for Python 3.8 to be able to do that.